test/test_flash_attn.py

```diff
@@ -0,0 +1,1695 @@
+        assert (dk - dk_ref).abs().max().item() <= mult * (dk_pt - dk_ref).abs().max().item() + 2e-4
+        assert (dv - dv_ref).abs().max().item() <= mult * (dv_pt - dv_ref).abs().max().item() + 2e-4
+    
+    print()
+
+
+@pytest.mark.parametrize("dtype", [torch.float16])
+@pytest.mark.parametrize("causal", [False, True])
+# @pytest.mark.parametrize('causal', [False])
+# @pytest.mark.parametrize("d", [32, 40, 59, 64, 80, 96, 111, 128, 160, 192, 224, 256])
+# @pytest.mark.parametrize('d', [32, 56, 64, 80, 96, 128])
+@pytest.mark.parametrize("d", [32, 64, 96, 128, 160, 192])
+# @pytest.mark.parametrize('d', [128])
+@pytest.mark.parametrize(
+    "seqlen_q,seqlen_k",
+    [
+        (1, 239),
+        (239, 1),
+        (3, 799),
+        (799, 3),
+        (1024, 128),
+        (97, 97),
+        (128, 128),
+        (200, 200),
+        (256, 256),
+        (257, 257),
+        (384, 384),
+        (512, 512),
+        (768, 768),
+        (1024, 1024),
+    ],
+)
+# @pytest.mark.parametrize('seqlen_q,seqlen_k', [(128, 128)])
+# @pytest.mark.parametrize("dropout_p", [0.0, 0.17])
+@pytest.mark.parametrize("dropout_p", [0.0])
+def test_flash_attn_race_condition(seqlen_q, seqlen_k, d, dropout_p, causal, dtype):
+    print(f"Running test with seqlen_q={seqlen_q}, seqlen_k={seqlen_k}, d={d}, dropout_p={dropout_p}, causal={causal}, dtype={dtype}")
+    device = xm.xla_device()
+    # set seed
+    torch.random.manual_seed(0)
+    batch_size = 60  # Sometimes we need large batch size for the race conditions to trigger
+    nheads = 4
+    q = torch.randn(batch_size, seqlen_q, nheads, d, dtype=dtype, requires_grad=True).to(device)
+    k = torch.randn(batch_size, seqlen_k, nheads, d, dtype=dtype, requires_grad=True).to(device)
+    v = torch.randn(batch_size, seqlen_k, nheads, d, dtype=dtype, requires_grad=True).to(device)
+    torch.random.manual_seed(42)
+    out0, lse0, _ = xf.flash_attn(q, k, v, dropout_rate=dropout_p, is_causal=causal, return_softmax=True)
+    g = torch.randn_like(out0)
+    if (d <= MAX_HEADDIM_SM8x or (d > 224 and dropout_p == 0)) or (is_sm80 or is_sm90):
+        (
+            dq0,
+            dk0,
+            dv0,
+        ) = torch.autograd.grad(out0, (q, k, v), g)
+        # Numerical error if we just do any arithmetic on dq
+        dq_atol = 2 * ((dq0 + 0.3 - 0.3) - dq0).abs().max().item()
+
+    for i in range(250):
+        torch.random.manual_seed(42)
+        out, lse, _ = xf.flash_attn(q, k, v, dropout_rate=dropout_p, is_causal=causal, return_softmax=True)
+        assert torch.equal(out, out0)
+        assert torch.equal(lse, lse0)
+
+        if (d <= MAX_HEADDIM_SM8x or (d > 224 and dropout_p == 0)) or (is_sm80 or is_sm90):
+            (
+                dq,
+                dk,
+                dv,
+            ) = torch.autograd.grad(out, (q, k, v), g)
+            dq_equal = torch.allclose(dq, dq0, atol=dq_atol)
```

Comment:
add unittest to compare xla.flash_attn and torch.nn.functional.scaled_dot_product_attention